### PR TITLE
[WIP] lightning glue example uses nlp package

### LINKED ISSUE #3692: How to use Huggingface pytorch bert to generate the prediction TSV file from the test set of a GLUE task?
# ❓ Questions & Help

## Details

Hello folks! Can you provide some simple example of using pytorch bert to generate the prediction TSV file from the test set of a GLUE task (such as MRPC) based on the fine-tuned model so that I can submit the prediction TSV file for each GLUE task to GLUE leaderboard? Thank you very much.

**A link to original question on Stack Overflow**:

### LINKED ISSUE #3925: Using the default trainer args
Hey guys!

Just looked through the transformer_base... why not also allow any of the trainer args to be used?

I imagine the best set up for you guys is:
- add default trainer args (since people can look into the lightning docs for details on this)
- add hf specific args (nlp related or whatever you guys need)
- add model specific args

But instead it looks like you only use a subset of the args
https://github.com/huggingface/transformers/blob/dd9d483d03962fea127f59661f3ae6156e7a91d2/examples/transformer_base.py#L275

Something like this should work:
```python
parser = ArgumentParser()

# enable all trainer args
parser = Trainer.add_argparse_args(parser)

# add the HF args
parser.add_argument(--some_hf_specific_thing, ...)

# add the model args
parser = GoodGAN.add_model_specific_args(parser)

# cook them all up :)
args = parser.parse_args()
```

Check here for mode details:
https://pytorch-lightning.readthedocs.io/en/stable/hyperparameters.html#multiple-lightning-modules

@srush @nateraw

### PR DESCRIPTION
**Pull request by nateraw**
Main goal here was to use `nlp` library to load datasets instead of custom scripts. I've made some changes to the GLUE example, and will reflect those changes elsewhere if the patterns used seem reasonable. Feedback would be greatly appreciated.

**So far, this PR:**
- Uses `nlp` library instead of manual processing scripts to download/process the benchmark datasets (Ref #4494)
- Uses lightning `Trainer` default arguments instead of previous custom logic. See[ list of available args](https://pytorch-lightning.readthedocs.io/en/latest/trainer.html#trainer-class) in their documentation. Resolves #3925 
- Generates submission files on test sets (partially resolves #3692)
- Fixes bug mentioned in #4214
- Pins lightning version to latest stable release, as master branch is a little to volatile for my taste.
- upgrades to pl=.76

**TODOs**
- Validate multi-gpu logic is up to date with lightning base practices
- Validate TPU logic is up to date with lightning best practices
- See if there's a better way to save dataset with `nlp` lib directly instead of current `torch.save` logic
- Run exhaustive test over model types over each dataset
- Optionally generate larger output table reporting benchmarks across models

**Note on Trainer args**

The argument parser will now accept any kwargs from the `Trainer` class's init function. For example, to load a checkpoint and run predictions to get submission files, you could run something like this to get a submission file at `./submissions/mrpc_submission.csv`:

```bash
export TASK=mrpc
export DATA_DIR=./cached_glue_data
export MAX_LENGTH=128
export LEARNING_RATE=2e-5
export BERT_MODEL=bert-base-cased
export BATCH_SIZE=32
export NUM_EPOCHS=3
export SEED=2
export GPUS=1
export NUM_WORKERS=4

# Add parent directory to python path to access lightning_base.py
export PYTHONPATH="../":"${PYTHONPATH}"

python3 -i run_pl_glue.py \
    --model_name_or_path $BERT_MODEL \
    --task $TASK \
    --data_dir $DATA_DIR \
    --max_seq_length  $MAX_LENGTH \
    --max_epochs $NUM_EPOCHS \
    --learning_rate $LEARNING_RATE \
    --seed $SEED \
    --gpus $GPUS \
    --num_workers $NUM_WORKERS \
    --train_batch_size $BATCH_SIZE \
    --resume_from_checkpoint ./lightning_logs/version_0/checkpoints/epoch=1.ckpt \
    --output_dir ./submissions/
    --do_predict
```

CC: @srush @williamFalcon

### CODE REVIEW

File: examples/text-classification/run_pl_glue.py [hunk 1]
```diff
@@ -1,115 +1,213 @@
...
-        logger.info("Loading features from cached file %s", cached_features_file)
-        features = torch.load(cached_features_file)
-        all_input_ids = torch.tensor([f.input_ids for f in features], dtype=torch.long)
-        all_attention_mask = torch.tensor([f.attention_mask for f in features], dtype=torch.long)
-        all_token_type_ids = torch.tensor([f.token_type_ids for f in features], dtype=torch.long)
-        if self.hparams.glue_output_mode == "classification":
-            all_labels = torch.tensor([f.label for f in features], dtype=torch.long)
-        elif self.hparams.glue_output_mode == "regression":
-            all_labels = torch.tensor([f.label for f in features], dtype=torch.float)
+            # TODO - check if this is still right or if it needs updating
```
Comment (julien-c) [inline examples/text-classification/run_pl_glue.py:101]: you should not need this anymore because `tokenizer.batch_encode_plus` would return just the relevant attributes according to the tokenizer's (model's) type.

Comment (nateraw) [reply examples/text-classification/run_pl_glue.py:101]: I'm using that `cols_to_keep` variable to filter the dataset down to just the kwargs that would be passed to a model's call function. In this case, do you think I still need it? Is there a more elegant way to write it instead?

Comment (julien-c) [reply examples/text-classification/run_pl_glue.py:101]: I haven't run your code so I'm not sure. I'm just saying that `tokenizer.batch_encode_plus` should now not return  `token_type_ids` if it's not used in the model afterwards.

File: examples/lightning_base.py [hunk 2]
```diff
@@ -139,7 +125,9 @@ def _feature_file(self, mode):
         )
 
     @staticmethod
-    def add_model_specific_args(parser, root_dir):
+    def add_model_specific_args(parent_parser):
```
Comment (nateraw) [inline examples/lightning_base.py:128]: Will probably need to parse temp args here to import `xm`. Otherwise the style checks will continue to fail on undefined variable. @williamFalcon - thoughts? Is `xm` still needed here for TPU?

File: examples/lightning_base.py [hunk 3]
```diff
@@ -96,25 +93,14 @@ def optimizer_step(self, epoch, batch_idx, optimizer, optimizer_idx, second_orde
         optimizer.zero_grad()
         self.lr_scheduler.step()
 
-    def get_tqdm_dict(self):
-        avg_loss = getattr(self.trainer, "avg_loss", 0.0)
-        tqdm_dict = {"loss": "{:.3f}".format(avg_loss), "lr": self.lr_scheduler.get_last_lr()[-1]}
-        return tqdm_dict
-
-    def test_step(self, batch, batch_nb):
```
Comment (nateraw) [inline examples/lightning_base.py:104]: removing `test_step` and `test_end` will likely break the other lightning examples. Will have to update their code accordingly.

File: examples/lightning_base.py [hunk 4]
```diff
@@ -96,25 +93,14 @@ def optimizer_step(self, epoch, batch_idx, optimizer, optimizer_idx, second_orde
         optimizer.zero_grad()
         self.lr_scheduler.step()
```
Comment (nateraw) [inline examples/lightning_base.py:94]: Do we still need to manually call `self.lr_scheduler` or can we have lightning handle this for us?

File: examples/lightning_base.py [hunk 5]
```diff
@@ -96,25 +93,14 @@ def optimizer_step(self, epoch, batch_idx, optimizer, optimizer_idx, second_orde
...
-
     def train_dataloader(self):
         train_batch_size = self.hparams.train_batch_size
         dataloader = self.load_dataset("train", train_batch_size)
 
         t_total = (
-            (len(dataloader.dataset) // (train_batch_size * max(1, self.hparams.n_gpu)))
-            // self.hparams.gradient_accumulation_steps
-            * float(self.hparams.num_train_epochs)
+            (len(dataloader.dataset) // (train_batch_size * max(1, self.hparams.gpus)))
```
Comment (nateraw) [inline examples/lightning_base.py:101]: Does this take into consideration TPU cores? Is that something that it needs to do?

File: examples/lightning_base.py [hunk 6]
```diff
@@ -123,7 +109,7 @@ def train_dataloader(self):
         return dataloader
 
     def val_dataloader(self):
-        return self.load_dataset("dev", self.hparams.eval_batch_size)
+        return self.load_dataset("validation", self.hparams.eval_batch_size)
```
Comment (nateraw) [inline examples/lightning_base.py:112]: This will break other lightning examples too

File: examples/text-classification/run_pl_glue.py (hunk 1 above)
Comment (nateraw) [reply examples/text-classification/run_pl_glue.py:101]: We can probably just use the map function's remove columns parameter and get rid of the `cols_to_keep` stuff...I'll take a look

**Comment by nateraw** [comment]
@julien-c I'm not sure why my checks for isort aren't passing on run_pl_glue.py. Could it be because I imported `pyarrow` which is not listed in third party within `setup.cfg`?

**Comment by williamFalcon** [comment]
lmk when you’re ready for me to give this a look over :)

one major feature we added is the option to not rely on hparams, they are of course backward compatible but now you have the option to instead pass all the args to init directly and we’ll still save the correct stuff in the checkpoint

**Comment by nateraw** [comment]
@sshleifer can you advise on the style check issue when you get the chance, please?

File: examples/text-classification/run_pl.sh [hunk 7]
```diff
@@ -1,36 +1,32 @@
...
---model_name_or_path $BERT_MODEL \
---output_dir $OUTPUT_DIR \
---max_seq_length  $MAX_LENGTH \
---learning_rate $LEARNING_RATE \
---num_train_epochs $NUM_EPOCHS \
---train_batch_size $BATCH_SIZE \
---seed $SEED \
---do_train \
---do_predict
+python3 -i run_pl_glue.py \
```
Comment (sshleifer) [inline examples/text-classification/run_pl.sh:20]: what does `-i` do? 
Can it solve all of our pythonpath problems!?

File: examples/text-classification/run_pl_glue.py [hunk 8]
```diff
@@ -1,115 +1,205 @@
...
 
         super().__init__(hparams, num_labels, self.mode)
 
     def forward(self, **inputs):
         return self.model(**inputs)
 
     def training_step(self, batch, batch_idx):
-        inputs = {"input_ids": batch[0], "attention_mask": batch[1], "labels": batch[3]}
-
-        if self.config.model_type != "distilbert":
```
Comment (sshleifer) [inline examples/text-classification/run_pl_glue.py:38]: How do we know it's safe to remove this?

File: examples/text-classification/run_pl_glue.py [hunk 9]
```diff
@@ -1,115 +1,205 @@
...
+        cached_dataset_file = self._feature_file(self.hparams.task)
+
+        if os.path.exists(cached_dataset_file) and not self.hparams.overwrite_cache:
+            logger.info("Loading from cached dataloader file ", cached_dataset_file)
+            self.dataset = torch.load(cached_dataset_file)
+        else:
+            logger.info("No cache file found. Creating one at", cached_dataset_file)
+
+            # Make the directory to hold cached dataset if it doesn't already exist
+            os.makedirs(os.path.dirname(cached_dataset_file), exist_ok=True)
```
Comment (sshleifer) [inline examples/text-classification/run_pl_glue.py:70]:
```python
from Pathlib import Path
Path(cached_dataset_file).parent.mkdir(exist_ok=True)
```

File: examples/text-classification/run_pl_glue.py [hunk 10]
```diff
@@ -1,115 +1,205 @@
...
+                Args:
+                    example_batch (Dict[List]): Dict with feature names as keys and batch values as values
+
+                Returns:
+                    Dict[List]: The desired updates to the batch (keys are feature names, values are batch values)
+                """
+
+                # Either encode single sentence or sentence pairs
+                if len(text_fields) > 1:
+                    texts_or_text_pairs = list(zip(example_batch[text_fields[0]], example_batch[text_fields[1]]))
```
Comment (sshleifer) [inline examples/text-classification/run_pl_glue.py:87]: is call to `list` needed?

File: examples/text-classification/run_pl_glue.py [hunk 11]
```diff
@@ -1,115 +1,205 @@
...
-                )
-                logger.info("Saving features into cached file %s", cached_features_file)
-                torch.save(features, cached_features_file)
-
-    def load_dataset(self, mode, batch_size):
-        "Load datasets. Called after prepare data."
 
-        # We test on dev set to compare to benchmarks without having to submit to GLUE server
-        mode = "dev" if mode == "test" else mode
+                # Rename 'label' to 'labels' so we can directly unpack batches into model call function
```
Comment (sshleifer) [inline examples/text-classification/run_pl_glue.py:96]: nice

File: examples/text-classification/run_pl_glue.py [hunk 12]
```diff
@@ -1,115 +1,205 @@
...
+        if self.hparams.task == "mnli":
+            logs = {}
+            for i, split in enumerate(["matched", "mismatched"]):
+                split_logs, _ = self._eval_end(outputs[i])
+                # Convert 'validation_loss' to 'validation_loss_matched' or 'validation_loss_mismatched'
+                logs.update({k + f"_{split}": v for k, v in split_logs.items()})
 
-        if self.config.model_type != "distilbert":
-            inputs["token_type_ids"] = batch[2] if self.config.model_type in ["bert", "xlnet", "albert"] else None
+            # HACK - pytorch lightning is looking for 'val_loss' as a key in returned dict.
```
Comment (sshleifer) [inline examples/text-classification/run_pl_glue.py:163]: thx for the comment

File: examples/text-classification/run_pl_glue.py [hunk 13]
```diff
@@ -118,78 +208,129 @@ def _eval_end(self, outputs):
...
+
+def main(args):
+
+    # Init Model
     model = GLUETransformer(args)
-    trainer = generic_train(model, args)
 
-    # Optionally, predict on dev set and write to output_dir
+    # Init Trainer
+    trainer = Trainer.from_argparse_args(args)
```
Comment (sshleifer) [inline examples/text-classification/run_pl_glue.py:283]: can we use `generic_train` or no?

File: examples/text-classification/run_pl_glue.py [hunk 14]
```diff
@@ -118,78 +208,129 @@ def _eval_end(self, outputs):
...
+    # Run training and reload best model from checkpoint
+    if args.do_train:
+
+        # Fit the model
+        trainer.fit(model)
+
+        # Reload best model from current experiment run's checkpoint directory
+        experiment_ckpt_dir = model.trainer.weights_save_path
+        checkpoints = list(sorted(glob.glob(os.path.join(experiment_ckpt_dir, "epoch=*.ckpt"), recursive=True)))
+        trainer.resume_from_checkpoint = checkpoints[-1]
```
Comment (sshleifer) [inline examples/text-classification/run_pl_glue.py:297]: can we test/assert that this change actually causes a new state dict to be loaded? or get some input from @williamFalcon ?

File: examples/text-classification/run_pl_glue.py [hunk 15]
```diff
@@ -118,78 +208,129 @@ def _eval_end(self, outputs):
...
+
+    # Predict on test split and write to submission files
     if args.do_predict:
-        checkpoints = list(sorted(glob.glob(os.path.join(args.output_dir, "checkpointepoch=*.ckpt"), recursive=True)))
-        model = model.load_from_checkpoint(checkpoints[-1])
+
+        # If we didn't train and and output directory is not supplied, raise an exception
+        if not args.do_train and not bool(args.output_dir):
+            raise RuntimeError("No output_dir is specified for writing results. Try setting --output_dir flag")
```
Comment (sshleifer) [inline examples/text-classification/run_pl_glue.py:304]: `--output_dir` should me required IMO

**Review by sshleifer — APPROVED** [review]
Nice. Would love to bump PL to .76 in examples/requirements.txt but that can be separate.

For style stuff, I have 
```
isort==4.3.21
black==19.10b0
```

**Review by sshleifer — COMMENTED** [review]
Nice. ~Would love to bump PL to .76 in examples/requirements.txt but that can be separate.~

For style stuff, I have 
```
isort==4.3.21
black==19.10b0
```

File: examples/text-classification/run_pl.sh (hunk 7 above)
Comment (nateraw) [reply examples/text-classification/run_pl.sh:20]: hah, unfortunately not. It just launches you into an interactive python interpreter when the code finishes running.

File: examples/text-classification/run_pl_glue.py (hunk 8 above)
Comment (nateraw) [reply examples/text-classification/run_pl_glue.py:38]: Its a little sneaky. I wrangled the correct keys/values in the `prepare_data` function.

File: examples/text-classification/run_pl_glue.py (hunk 9 above)
Comment (nateraw) [reply examples/text-classification/run_pl_glue.py:70]: I'm a huge fan of this idea. Kept it as `os` as the other examples all use that, but everyone knows `pathlib` is better. 😄

File: examples/text-classification/run_pl_glue.py (hunk 10 above)
Comment (nateraw) [reply examples/text-classification/run_pl_glue.py:87]: I'm actually not sure...will investigate. Nice catch!!

File: examples/text-classification/run_pl_glue.py (hunk 13 above)
Comment (nateraw) [reply examples/text-classification/run_pl_glue.py:283]: I got rid of it. Instead, we use the built in `Trainer` argparser (and the base model/subclass argparsers) to handle everything `generic_train` was doing before.

File: examples/text-classification/run_pl_glue.py (hunk 14 above)
Comment (nateraw) [reply examples/text-classification/run_pl_glue.py:297]: Very good point. I've never seen anybody do what I did here and I didn't test it myself...just kinda assumed when `trainer.test` or `trainer.fit` are called, it would look to see what is set for `trainer.resume_from_checkpoint` and load if need be.

**Comment by nateraw** [comment]
Adding note for myself...seems `pandas` has been removed from lightning's requirements.txt so we may need to add that to `examples/requirements.txt`.

File: examples/text-classification/run_pl_glue.py (hunk 10 above)
Comment (nateraw) [reply examples/text-classification/run_pl_glue.py:87]: confirmed not needed

File: examples/text-classification/run_pl_glue.py (hunk 14 above)
Comment (nateraw) [reply examples/text-classification/run_pl_glue.py:297]: confirmed it works

**Comment by nateraw** [comment]
Closing and will reopen cleaner one later.
